### PR TITLE
Enforce extension validity at parse time

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

This PR adds a `DistExtension` field to some of our distribution types, which requires that we validate that the file type is known and supported when parsing (rather than when attempting to unzip). It removes a bunch of extension parsing from the code too, in favor of doing it once upfront.

Closes https://github.com/astral-sh/uv/issues/5858.

### CODE REVIEW

**Comment by charliermarsh** [comment]
I think this is a good change but it's more aggressive. It means we might error if a registry serves distribution kinds that we don't support, even if we'd never read them.

**Comment by charliermarsh** [comment]
Oh nevermind, we wouldn't fail in that case due to some other gating.

File: crates/uv/tests/pip_install.rs [hunk 1]
```diff
@@ -5705,23 +5705,15 @@ fn tool_uv_sources() -> Result<()> {
...
-     + boltons==24.0.1.dev0 (from git+https://github.com/mahmoud/boltons@57fbaa9b673ed85b32458b31baeeae230520e4a0)
-     + colorama==0.4.6
-     + idna==3.6
-     + packaging==24.1.dev0 (from git+https://github.com/pypa/packaging@32deafe8668a2130a3366b98154914d188f3718e)
-     + poetry-editable==0.1.0 (from file://[TEMP_DIR]/poetry_editable)
-     + sniffio==1.3.1
-     + tqdm==4.66.0 (from https://files.pythonhosted.org/packages/a5/d6/502a859bac4ad5e274255576cd3e15ca273cdb91731bc39fb840dd422ee9/tqdm-4.66.0-py3-none-any.whl)
-     + urllib3==2.2.1 (from https://files.pythonhosted.org/packages/a2/73/a68704750a7679d0b6d3ad7aa8d4da8e14e151ae82e6fee774e6e0d05ec8/urllib3-2.2.1-py3-none-any.whl)
+    error: Failed to parse metadata from built wheel
+      Caused by: Expected direct URL dependency to include an extension: `https://example.org/does/not/exist`
```
Comment (zanieb) [inline crates/uv/tests/pip_install.rs:5714]: Can we say "file extension"? Why did this succeed previously?

**Review by zanieb — APPROVED** [review]
Basically down for this behavior.

File: crates/distribution-filename/src/extension.rs [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    PartialEq,
+    Eq,
+    PartialOrd,
+    Ord,
+    Hash,
+    Serialize,
+    Deserialize,
+    rkyv::Archive,
+    rkyv::Deserialize,
+    rkyv::Serialize,
```
Comment (charliermarsh) [inline crates/distribution-filename/src/extension.rs:41]: The existing definition was:
```rust
#[derive(
    Clone,
    Debug,
    PartialEq,
    Eq,
    Serialize,
    Deserialize,
    rkyv::Archive,
    rkyv::Deserialize,
    rkyv::Serialize,
)]
#[archive(check_bytes)]
#[archive_attr(derive(Debug))]
pub enum SourceDistExtension {
    Zip,
    TarGz,
    TarBz2,
}
```

I think this is safe without a cache bump? @BurntSushi?

**Review by konstin — APPROVED** [review]

**Review by BurntSushi — APPROVED** [review]
Yes, I like this!

**Comment by charliermarsh** [comment]
I bumped the rkyv cache, to be conservative.

File: crates/distribution-filename/src/extension.rs [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+            }
+            "xz" if path.as_ref().file_stem().is_some_and(|stem| {
+                Path::new(stem)
+                    .extension()
+                    .is_some_and(|ext| ext.eq_ignore_ascii_case("tar"))
+            }) =>
+            {
+                Ok(Self::TarXz)
+            }
+            "zst"
```
[the remark this answers is not in the document]
Comment (charliermarsh) [reply crates/distribution-filename/src/extension.rs]: Thanks, that's wrong!

**Pull request merged**